### PR TITLE
use most appropriate time information when routing between states

### LINKED ISSUE #900: Map-Matching - Error Rejecting High Speed Paths
Paths in map-matching are not being rejected based on estimated time vs. measurement time due to interpolated points (points close to one another).  Several examples have the first time followed by some measurements at basically the same position (interpolated) - but the map-matching route rejection uses the time difference between the first point and the next non-interpolated point even though the vehicle was stationary for some period of time leading to a higher max_cost. This diagram perhaps helps (s is a real measurement and i is an interpolated measurement)

...            .       .      ...               ...              Input measurements 
sii            s     s      sii               sis            Current
iis            s     s      sis              sis             Proposed

(... indicates measurements within the interpolation distance, s indicates a "real" measurement and i indicates an interpolated measurement)
With the proposed change the time between the first 2 real non-interpolated measurements will be shorter (from the last of the nearby measurements rather than the first).  Intermediate measurements need to have an "entering" real measurement and a "leaving" real measurement in the grouping of nearby input measurements.

@ptpt @kevinkreiser

### PR DESCRIPTION
**Pull request by kevinkreiser**
this fixes #900 

the main idea here is found in a single large comment, noted in the code. currently its failing the randomized map matching test so im trying to figure out why but it does fix the issue raised in #900.

### CODE REVIEW

File: src/meili/map_matcher.cc [hunk 1]
```diff
@@ -277,30 +277,36 @@ MapMatcher::OfflineMatch(
...
-    } else {
+    }// This one is so close to the last one we made state for that we will just interpolate it
+    else {
       interpolated_measurements[time].push_back(*measurement);
+      // If this interpolated point had time information we want to use that when the route
+      // leaves latest_match_measurement so we get accurate timing for the route between
+      // points. This protects against interpolated points being close in distance but far
+      // apart in time. We could instead not interpolating both on distance and on time
+      // but then this may lead to issues where the route doubles back on itself because
+      // the points are so close together and we dont want that
```
Comment (kevinkreiser) [inline src/meili/map_matcher.cc]: above the the description of what is going on here

**Comment by dnesbitt61** [comment]
:shipit:

**Comment by ptpt** [comment]
I think it would be better to put leave times in `MapMatching` to keep measurements immutable.
```c++
void SetMeasurementLeaveTime(StateId::Time time, double leave_time)
{ leave_times_[time] = leave_time; }

private:
std::vector<double> leave_times_;
```

**Comment by kevinkreiser** [comment]
@ptpt i had it that way and changed it! i can change it back but actually im looking at something else right now. specifically this change has bad results when someone is traveling slow and sends location at a high frequency. so like a car driving 5mph but giving updates every second. the tests exposed this problem with this idea so i'm working on something else to solve this.

**Comment by kevinkreiser** [comment]
@ptpt i changed it back can you have a look

File: src/meili/map_matcher.cc [hunk 2]
```diff
@@ -309,30 +309,45 @@ MapMatcher::OfflineMatch(
...
+    // Always match the last measurement and if its far enough away
+    if (sq_interpolation_distance < sq_distance || std::next(m) == measurements.end()) {
+      // If there were interpolated points between these two points with time information
+      if(interpolated_epoch_time != -1) {
+        // Project the last interpolated point onto the line between the two match points
+        auto p = interpolated[time].back().lnglat().Project(last->lnglat(), m->lnglat());
+        // If its significantly closer to the previous match point then it looks like the trace lingered
+        // so we use the time information of the last interpolation point as the actual time they started
+        // traveling towards the next match point which will help us determine what paths are really likely
+        if(p.Distance(last->lnglat())/last->lnglat().Distance(m->lnglat()) < .2f)
```
Comment (kevinkreiser) [inline src/meili/map_matcher.cc]: this `.2f` is a magic number. i can make it a constant or we can make it configurable, what do @ptpt and @dnesbitt61 think?

**Comment by ptpt** [comment]
Need to clear `leave_times_` in `MapMatching::Clear()` too.

File: src/meili/map_matching.cc [hunk 3]
```diff
@@ -172,8 +158,8 @@ MapMatching::TransitionCost(const StateId& lhs, const StateId& rhs) const
   const auto& right = state(rhs);
   const auto& left_measurement = measurement(left);
   const auto& right_measurement = measurement(right);
-  const auto gc_dist = GreatCircleDistance(left_measurement, right_measurement);
-  const auto clk_dist = ClockDistance(left_measurement, right_measurement);
+  const auto gc_dist = left_measurement.lnglat().Distance(right_measurement.lnglat());
+  const auto clk_dist = right_measurement.epoch_time() - leave_times_[lhs.time()];
```
Comment (ptpt) [inline src/meili/map_matching.cc]: It seems that we haven't consider the case that `right_measurement.epoch_time() == -1` or `leave_times_[lhs.time()] == -1`. Won't it lead to large clk_dist?

Comment (kevinkreiser) [reply src/meili/map_matching.cc]: i was noticing that myself and thought that i handled that somewhere but its slipped my mind. ill go back and look at the original pr that added support for the `time_factor`

Comment (kevinkreiser) [reply src/meili/map_matching.cc]: ok ive added a saftey check, the only way this problem could happen is if someone gives some measurements with times and some without or gives invalid times as input but anyway it should be safe now

**Comment by ptpt** [comment]
Looks good to me.

A minor issue is that it seems we don't have a consistent to define the case that timestamp doesn't exits. In measurment, it's `-1` (which I assume measurements with epoch < 0 doesn't have timestamp), but in find_shortest_path it is `max_time == 0` and in ClockDistance it is `right.epoch_time() > 0 && left.epoch_time() >= 0`.

Can we define:  `timestamp < 0 => invalid, otherwise it's valid`

**Comment by kevinkreiser** [comment]
@ptpt i've made them all use the same comparison. i could pull this out to a method somewhere if that is desired but i think its as you wanted for now

**Pull request merged**